### PR TITLE
Remove microseconds from metadata API Signed.expires

### LINKED ISSUE #1678: metadata api: Signed.expires string representation should not contain microseconds
**Description of issue or feature request**:

 In `signed._common_fields_to_dict` we convert the value of the datetime object in `Signed.expires` to a date-time string.

From the [TUF spec](https://theupdateframework.github.io/specification/latest/#date-time):
> [...] The expected format of the combined date and time string is "YYYY-MM-DDTHH:MM:SSZ". [...] An example date-time string is "1985-10-21T01:21:00Z".

If the original datetime object contained microseconds, the date-time string representation of `Signed.expires` will too.

**Reproducer:**
```python
from datetime import datetime
from tuf.api.metadata import Timestamp, MetaFile
Timestamp(1, "1.0.19", datetime(2021, 11, 16, 12, 0, 0, 123456), MetaFile(1)).to_dict()["expires"]
```

**Current output**:

```
'2021-11-16T12:00:00.123456Z'
```

**Expected behavior**:

```
'2021-11-16T12:00:00Z'
```

**Proposed fix 1**:
```diff
diff --git a/tuf/api/metadata.py b/tuf/api/metadata.py
index 25f14fe7..cdbd8595 100644
--- a/tuf/api/metadata.py
+++ b/tuf/api/metadata.py
@@ -480,7 +480,7 @@ class Signed(metaclass=abc.ABCMeta):
             "_type": self._type,
             "version": self.version,
             "spec_version": self.spec_version,
-            "expires": self.expires.isoformat() + "Z",
+            "expires": self.expires.replace(microsecond=0).isoformat() + "Z",
             **self.unrecognized_fields,
         }
```

**Proposed fix 2**:
```diff
diff --git a/tuf/api/metadata.py b/tuf/api/metadata.py
index 25f14fe7..20365cfe 100644
--- a/tuf/api/metadata.py
+++ b/tuf/api/metadata.py
@@ -480,7 +480,7 @@ class Signed(metaclass=abc.ABCMeta):
             "_type": self._type,
             "version": self.version,
             "spec_version": self.spec_version,
-            "expires": self.expires.isoformat() + "Z",
+            "expires": datetime.strftime(self.expires, "%Y-%m-%dT%H:%M:%SZ"),
             **self.unrecognized_fields,
         }
```

### PR DESCRIPTION
**Pull request by ivanayov**
This change removes microseconds from expiry in order to fit TUF
specification

Fixes #1678

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [n/a] Tests have been added for the bug fix or new feature
- [n/a] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by jku** [comment]
We'll want think this through for all the different cases: deserialization, users calling the api , serialization... it's a little detail but could make life difficult in the future
* I'd want to see docstrings for public API at least
* is it correct that we accept non-spec-compliant dates on deserialize?
* is it correct that we change the date to compliant on deserialize?

**Comment by ivanayov** [comment]
@jku 
> is it correct that we accept non-spec-compliant dates on deserialize?

It think that spec-compliant should mean spec-compliant with no exception, so if metadata is not in the correct form, we should throw an error.

> is it correct that we change the date to compliant on deserialize?

I think it's fine, when it comes to preventing a bug, coming not from the metadata, but from the implementation (and in this case - python specific)   as we guarantee that it won't break the deserialisation. But we should guarantee that the metadata is correct at the same time. 
May be I'd need to add a check, which is now missing.

**Comment by lukpueh** [comment]
I agree with @jku that we need to be careful here. This is what I think about the different cases:

* deserialization:
  The deserializer shouldn't accept non-spec compliant metadata, regardless of its importance. If we start modifying the semantics of inputs during deserialization, we'll likely end up with determinism problems (goal: `data == serialize(deserialize(data)`)

* Object API:
  Ideally, we don't accept non-sepc compliant inputs here either. But in reality this is hard to enforce, because properties can be set in different ways. So I think it is easier and okay to only enforce spec compliance on the de/serialization boundaries.

* Serialization:
  Under above assumptions, i.e. that we don't accept non-spec compliant metadata in deserialization, and that some properties that were modified during object life-time contain additional (unneeded) information, certain transformations should be fine, e.g. removing unnecessary microseconds. We need to find a balance between expected behaviour and usability.

**Comment by jku** [comment]
I think I lean in the same direction (EDIT: I wrote this before seeing lukas' last message :) )
* spec defines the date format exactly: we should only accept that in deserialization?
* constructing a new object should probably accept microseconds and just remove them
* deserialization and constructing new object currently use same code path (in the constructor)

so possibly 
* from_dict() should be strict, and should fail if format is not exactly as defined in spec
* but the constructor should use set_expiry() as your PR does: so constructing new object does not fail on microseconds ...

I'm not 100% sure of this yet, but that's my thinking...

**Comment by lukpueh** [comment]
I agree with the overall direction...

> from_dict() should be strict, and should fail if format is not exactly as defined in spec

This makes sense, and we might need to revise existing data validation in `from_dict/__init__`-methods, because I saw that some is performed in one and some in the other.

> but the constructor should use set_expiry() as your PR does: so constructing new object does not fail on microseconds

I guess we can do that, but we need to to be careful that there's no unexpected behavior, above all upon serialization, if the expires field is modified outside of the constructor

**Comment by jku** [comment]
> but we need to to be careful that there's no unexpected behavior, above all upon serialization, if the expires field is modified outside of the constructor

Yes, and should consider buggy code as well (obviously we'd never write bugs but just theoretically speaking): we should ensure we rather fail than persist metadata that is invalid. #1696 has some discussion on this

**Comment by ivanayov** [comment]
In the latest update I stick to:
* `__ini__()` uses `set_expiry()`, so that we guarantee microseconds are not added on runtime
* a format verification for `"expires"` in `_common_fields_from_dict()` 
* no format changes (which I pushed earlier) in `_common_fields_to_dict()`

File: tuf/api/metadata.py
Comment (kairoaraujo) [inline tuf/api/metadata.py]: There is a white trailing space at the end of this line.

**Review by kairoaraujo — COMMENTED** [review]
To test your code (lazily), I used the ``tests/metadata_api_tests/test_metadata_serialization.py`` and I added an invalid expire date entry to this data set https://github.com/theupdateframework/python-tuf/blob/d3b877b38ae0e73ff808003eb4321c6c5749ad1d/tests/test_metadata_serialization.py#L169

```python
        "invalid expire date": '{"_type": "root", "spec_version": "1.0.0", "version": 1, \
            "expires": "2021-11-16T12:00:00.123456Z", "consistent_snapshot": false, \
            "keys": { \
                "keyid1" : {"keytype": "rsa", "scheme": "rsassa-pss-sha256", "keyval": {"public": "foo"}}, \
                "keyid2" : {"keytype": "ed25519", "scheme": "ed25519", "keyval": {"public": "bar"}}}, \
            "roles": { \
                "root": {"keyids": ["keyid1"], "threshold": 1}, \
                "timestamp": {"keyids": ["keyid2"], "threshold": 1}, \
                "targets": {"keyids": ["keyid1"], "threshold": 1}, \
                "snapshot": {"keyids": ["keyid2"], "threshold": 1}} \
            }',
```

It gave me the expected raise:
```python
  File "/Users/kdearaujo/devel/TUF/python-tuf/tuf/api/metadata.py", line 476, in _common_fields_from_dict
    raise exceptions.InvalidConfigurationError(
tuf.exceptions.InvalidConfigurationError: 
            Role root should have "expires" metadata of the format 
            "yyyy-mm-ddThh:mm:ssZ" but got 2021-11-16T12:00:00.123456Z instead
```

But I am not sure if the right to a place to add the test or even we want to test it there. I noticed that we don't test these custom exceptions in ``tests_metadata_serialization.py``

Maybe some maintainer could comment :)

File: tuf/api/metadata.py
Comment (ivanayov) [reply tuf/api/metadata.py]: @kairoaraujo Yep, it's on purpose, as this is printed on a single line.

Comment (ivanayov) [reply tuf/api/metadata.py]: Regarding the tests - thank you for the notice! I was also thinking of adding one, but was waiting for feedback first to see if we agree on this approach.
`test_metadata_serialization.py` looks a good place to me. May be we should sync with @MVrachev as he's also working on validation.

Comment (kairoaraujo) [reply tuf/api/metadata.py]: My suggestion:

```python
        except ValueError:
            raise exceptions.InvalidConfigurationError(
                f"Role {_type} should have 'expires' metadata of the format "
                + f"'yyyy-mm-ddThh:mm:ssZ' but got {expires_str} instead"
            ) from None
```

Comment (lukpueh) [reply tuf/api/metadata.py]: +1 fo Kairo's suggestion. I'd generally not add newlines inside error/log messages unless they help readability of the message (not of the code).

One tiny addition, the `+` is not required for string concatenation inside braces.

Comment (lukpueh) [reply tuf/api/metadata.py]: Two more comments about the error:
- `InvalidConfigurationError` does not seem like the right exception type. I'd probably just re-raise the `ValueError` for now (see #1131)
- nit: I think the correct ISO8601 spelling would use upper case for the date part, i.e. "YYYY-MM-DDThh:mm:ssZ". But it seems that the [TUF spec](https://theupdateframework.github.io/specification/latest/#date-time) uses all upper case, i.e.  "YYYY-MM-DDTHH:MM:SSZ". I suggest you use either one of those, but I won't insist. :)

**Comment by lukpueh** [comment]
Thanks for working on this @ivanayov. I think it's a fair approach for the first two use cases discussed above. We still need to figure out what to do for the 3rd use case, that is do we allow serialisation of invalid metadata. But maybe we don't have to do this here. Would you mind adding a test for what you have so far?

**Comment by ivanayov** [comment]
@lukpueh @kairoaraujo Just pushed an update with latest comments.

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: I can see you've discussed this already (and I agree with using just a plain ValueError)... but what's wrong with the original ValueError thrown by strptime? I don't see an advantage to adding this code.

File: tests/test_api.py
Comment (jku) [inline tests/test_api.py]: I don't think this test is necessary -- the serialization test seems good enough to me (or at least we should be able to explain why we don't add hundreds of other tests like this for invalid json)

File: tests/test_metadata_serialization.py
Comment (jku) [inline tests/test_metadata_serialization.py]: some non-important nitpicks here:
* name doesn't say why it's invalid, why not say "invalid expiry with microsecs" ?
* keys could be an empty dict (and the roles keyids lists could be empty) to make this test data smaller

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: public method without docstring

There are also other methods dealing with expiry already in this class. This method should be grouped with them I think

**Review by jku — CHANGES_REQUESTED** [review]
I left some comments, I think the core of this PR is the new API (`set_expiry()`) which allows both construction of new metadata and modifications to existing expiry to be done safely: I think this is the correct direction but new API needs to to be polished (include API documentation and so on).

The big question is, do we need to solve the related #1727 at the same time (at least on paper)?

**Comment by jku** [comment]
> We still need to figure out what to do for the 3rd use case, that is do we allow serialisation of invalid metadata

Martin opened #1696 about this relevant question in general: maybe we can continue there (the issue is big enough to maybe have a chat as well, to discuss some possible ways to deal with it)

**Comment by jku** [comment]
oh a new idea: since one of the problems here is that our expiry is a datetime but most datetimes are not valid expiries (because microseconds), how about we hide the variable and make `expiry` a property (where the setter is Ivanas function) -- that would effectively prevent a user from accidentally setting `signed.expiry` to an invalid value: something like `signed.expiry = utcnow() + relativedelta(day=3)` would now work correctly because it would use the property setter.

File: tuf/api/metadata.py
Comment (lukpueh) [reply tuf/api/metadata.py]: I had the same thought, but I guess there's merit in a contextualised error message... of course, there's also merit in 7 lines of code less. :)

Comment (ivanayov) [reply tuf/api/metadata.py]: I see your point (as I was thinking the same), but in the end thought that general ValueError doesn't give information on what's expected. We can always say that one may just check the spec, so I'm fine with removing the change if you both vote for that.

**Review by jku — APPROVED** [review]
This feels like the correct solution, thank you.
* API user just provides a datetime, we make sure it's valid (removing microseconds) 
* the Metadata object always matches the serialized output (since the microseconds are never stored in the object)
* type "enforcement" is about as good as it gets with python (static checks will catch wrong types and we use the datetime API in the setter so even without static checks things will break on type mismatches) 

Looks good to me, leaving open for @lukpueh to see as well

**Review by lukpueh — APPROVED** [review]
Seconding Jussi's assessment. Thanks for the clean fix, @ivanayov!

**Pull request merged**
